src/eigen/lobpcg_hyper_impl.jl

```diff
@@ -43,17 +43,83 @@
 vprintln(args...) = nothing
 
 using LinearAlgebra
-using BlockArrays # used for the `mortar` command which makes block matrices
+import Base: *
+include("../workarounds/gpu_arrays.jl")
 
-# when X or Y are BlockArrays, this makes the return value be a proper array (not a BlockArray)
-function array_mul(X::AbstractArray{T}, Y) where T
-    Z = Array{T}(undef, size(X, 1), size(Y, 2))
-    mul!(Z, X, Y)
+# For now, BlockMatrix can store arrays of different types (for example, an element 
+# of type views and one of type Matrix). Maybe for performance issues it should only
+# store arrays of the same type?
+
+struct BlockMatrix
+    blocks::Tuple
+    size::Tuple{Int64,Int64}
+end
+
+"""
+Build a BlockMatrix containing the given arrays, from left to right.
+This function will fail (for now) if:
+    -the arrays do not all have the same "height" (ie size[1] must match).
+"""
+function make_block_vector(arrays::AbstractArray...)
+    length(arrays) ==0 && error("Empty BlockMatrix is not currently implemented")
+    n_ref= size(arrays[1])[1]
+    m=0
+    for array in arrays
+        n_i, m_i = size(array)
+        n_ref != n_i && error("The given arrays do not have matching 'height': cannot build a BlockMatrix out of them.")
+        m += m_i
+    end
+    BlockMatrix(arrays, (n_ref,m))
+end
+
+
+"""
+Given A and B as two BlockMatrixs [A1, A2, A3], [B1, B2, B3] form the matrix
+A'B (which is not a BlockMatrix). block_overlap also has compatible versions with two Arrays. 
+block_overlap always compute some form of adjoint, ie the product A'*B.
+"""
+@views function block_overlap(A::BlockMatrix, B::BlockMatrix)
```

Comment:
Why is this function not covered by tests? Did you check you end up dispatching correctly?

Reply:
Also not too big a fan of having this function. Why not define
```julia
Base.:*(A::Adjoint{T, <: BlockMatrix}, B::BlockMatrix) where {T})
```
instead?

Reply:
At first I thought defining it this way would be clearer: as it turns out, it's just weird and confusing. I re-wrote the `block_overlap` functions, now we only overload `Base.*`.